flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java

```diff
@@ -406,6 +407,43 @@ public void onError(Throwable error) {
 		onFatalError(error);
 	}
 
+	@Override
+	public void onContainerStarted(ContainerId containerId, Map<String, ByteBuffer> allServiceResponse) {
+		log.info("The container {} started.", containerId);
+	}
+
+	@Override
+	public void onContainerStatusReceived(ContainerId containerId, ContainerStatus containerStatus) {
+	}
+
+	@Override
+	public void onContainerStopped(ContainerId containerId) {
+		log.info("The container {} is stopped.", containerId);
+	}
+
+	@Override
+	public void onStartContainerError(ContainerId containerId, Throwable t) {
+		log.error("Could not start TaskManager in container {}.", containerId, t);
+
+		// release the failed container
+		YarnWorkerNode yarnWorkerNode = workerNodeMap.remove(new ResourceID(containerId.toString()));
```

Comment:
Maybe it safer to check whether yarnWorkerNode is null.

Reply:
Will change.